### PR TITLE
633176: Handle queue deletion

### PR DESCRIPTION
**Pull request by rorytorneymf**
When a queue is deleted, the old `declaredQueues` list would still contain the deleted queue, so the next time `createStagingQueue` is called, it would see an entry in this list, think the queue already exists, and return without creating the queue. This would block work for this tenant forever (or until the Job Service Scheduled Executor is restarted and the `declaredQueues` is reset to empty again).

Replacing this list with a cache of all staging queue names, fetched from the RabbitMQ API.

https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=633176

### CODE REVIEW

File: worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreator.java [hunk 1]
```diff
@@ -101,7 +111,15 @@ void createStagingQueue(final Queue targetQueue, final String stagingQueueName)
             // Throw original exception
             throw ioException;
         }
+    }
 
-        declaredQueues.add(stagingQueueName);
+    private List<String> getPossiblyCachedStagingQueueNames()
+    {
+        return cachingQueuesApi.getApi()
```
Comment (andyreidz) [inline worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreator.java:118]: Need to verify the performance of this, will every call to getPossiblyCachedStagingQueueNames result in the cached JSON string being deserialised into Java objects?
The concern I mentioned, when we spoke last week, is that some worker process messages at 100/s and this deserialisation will be expensive.

File: worker-message-prioritization-rerouting/src/test/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreatorIT.java [hunk 2]
```diff
@@ -70,4 +70,74 @@ public void createStagingQueueTest() throws TimeoutException, IOException, Inter
...
+
+                // Delete the staging queue
+                channel.queueDelete(stagingQueueName);
+
+                // Verify the staging queue was deleted successfully
+                try {
+                    queuesApi.getApi().getQueue("/", stagingQueueName);
+                    Assert.fail("Expected a 404 when trying to get a queue that has been deleted");
+                } catch (final Exception e) {
+                    // Expected
```
Comment (andyreidz) [inline worker-message-prioritization-rerouting/src/test/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreatorIT.java:117]: Check the exception is the one you expect

File: worker-message-prioritization-rerouting/src/test/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreatorIT.java [hunk 3]
```diff
@@ -70,4 +70,74 @@ public void createStagingQueueTest() throws TimeoutException, IOException, Inter
...
+                // this because the staging queue is present inside the StageQueueCreator's HTTP cache. As such, the staging queue
+                // should not be created.
+                stagingQueueCreator.createStagingQueue(targetQueue, stagingQueueName);
+
+                // Verify the staging queue was not created
+                try {
+                    queuesApi.getApi().getQueue("/", stagingQueueName);
+                    Assert.fail("Expected a 404 when trying to get a queue that has been deleted");
+                } catch (final Exception e) {
+                    // Expected
```
Comment (andyreidz) [inline worker-message-prioritization-rerouting/src/test/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreatorIT.java:130]: Check the exception is what you expect

**Review by andyreidz — CHANGES_REQUESTED** [review]
Have a concern over repeated JSON deserialization.

File: worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreator.java (hunk 1 above)
Comment (rorytorneymf) [reply worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreator.java:118]: Yes the response is deserialized into Java objects, even when the response is served from the cache.

Should we just [go back to using the LoadingCache](https://github.com/WorkerFramework/worker-message-prioritization/pull/20/commits/a65785f4d21d2e140eee4fce5bc90471577f17ba) which doesn't have this problem?

Comment (andyreidz) [reply worker-message-prioritization-rerouting/src/main/java/com/github/workerframework/workermessageprioritization/rerouting/StagingQueueCreator.java:118]: If there isn't a way to have retrofit cache the deserialized list then perhaps try
https://www.baeldung.com/guava-memoizer

**Review by andyreidz — APPROVED** [review]

**Pull request merged**